mongo/mongologger/mongologger.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package mongologger // import "go.mongodb.org/mongo-driver/mongo/mongologger"
+
+import (
+	"os"
+	"strconv"
+)
+
+// Logger is an interface for loggers to implement
+type Logger interface {
+	Trace(message string, args ...Field)
+	Debug(message string, args ...Field)
+	Info(message string, args ...Field)
+	Notice(message string, args ...Field)
+	Warning(message string, args ...Field)
+	Error(message string, args ...Field)
+}
+
+// MongoLogger manages the mongo log messages sent to the logger
+type MongoLogger struct {
+	logger Logger
+
+	logFullCommands      bool
+	commandLevel         Level
+	connectionLevel      Level
+	sdamLevel            Level
+	serverSelectionLevel Level
+}
+
+// NewMongoLogger creates a new Mongologger to use with a mongo.Client. Log level defaults to Warning
+// for all components.
+func NewMongoLogger(opts ...*Options) (*MongoLogger, error) {
+	mlo := MergeOptions(opts...)
+
+	if mlo.Logger == nil {
+		dl := &defaultLogger{}
+
+		dl.writer = os.Stderr
+		var err error
+		if mlo.OutputFile != nil {
+			dl.outputFile = *mlo.OutputFile
+			dl.writer, err = os.Create(dl.outputFile)
+			if err != nil {
+				return nil, err
+			}
+		}
+		mlo.Logger = dl
+	}
+
+	ml := MongoLogger{
+		logger: mlo.Logger,
+
+		commandLevel:         Warning,
+		connectionLevel:      Warning,
+		sdamLevel:            Warning,
+		serverSelectionLevel: Warning,
+	}
+
+	if mlo.LogFullCommands != nil {
+		ml.logFullCommands = *mlo.LogFullCommands
+	}
+	if mlo.CommandLevel != nil {
+		ml.commandLevel = *mlo.CommandLevel
+	}
+	if mlo.ConnectionLevel != nil {
+		ml.connectionLevel = *mlo.ConnectionLevel
+	}
+	if mlo.SDAMLevel != nil {
+		ml.sdamLevel = *mlo.SDAMLevel
+	}
+	if mlo.ServerSelectionLevel != nil {
+		ml.serverSelectionLevel = *mlo.ServerSelectionLevel
+	}
+
+	return &ml, nil
+}
+
+// NewMongoLoggerFromEnv returns a mongologger set to the environment variables. If not logging environment variables
+// are set, returns nil
+func NewMongoLoggerFromEnv() (*MongoLogger, error) {
```

Comment:
This seems like a weird function to have in our public API because we don't actually expect anyone to call it besides the `mongo.Client` configuration code. The API is also a little weird because constructors don't usually return nil. If it's only called to construct a logger in the case where one isn't set, can it be a private function that we declare in the same package as where it's used?